### PR TITLE
Rustc metadata diagnostics

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
r? @jonathandturner

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
This is a redoing from #34970.

File: src/librustc_metadata/diagnostics.rs [hunk 1]
```diff
@@ -91,6 +91,185 @@ You need to link your code to the relevant crate in order to be able to use it
...
+    ...
+}
+
+#[macro_export]
+macro_rules! bring_beer {
+    ...
+}
+
+// In your crate:
+#[macro_use(get_tacos, bring_beer)] // It imports macros get_tacos and
```
Comment (jonathandturner) [inline src/librustc_metadata/diagnostics.rs]: We could probably use a better example than "beer and tacos" ;)

Comment (GuillaumeGomez) [inline src/librustc_metadata/diagnostics.rs]: You're hard on business. :-/

**Comment by GuillaumeGomez** [comment]
Beer became pimiento!

File: src/librustc_metadata/diagnostics.rs [hunk 2]
```diff
@@ -91,6 +91,185 @@ You need to link your code to the relevant crate in order to be able to use it
...
+    ...
+}
+
+#[macro_export]
+macro_rules! get_pimientos {
+    ...
+}
+
+// In your crate:
+#[macro_use(get_tacos, get_pimiento)] // It imports macros get_tacos and
```
Comment (jonathandturner) [inline src/librustc_metadata/diagnostics.rs]: Should this be `get_pimientos`?

File: src/librustc_metadata/diagnostics.rs [hunk 3]
```diff
@@ -91,6 +91,185 @@ You need to link your code to the relevant crate in order to be able to use it
...
+extern crate some_crate;              // bring_beer macros from some_crate.
```
Comment (jonathandturner) [inline src/librustc_metadata/diagnostics.rs]: We're still bringing beer...?

Comment (GuillaumeGomez) [inline src/librustc_metadata/diagnostics.rs]: Ah damn!

**Comment by GuillaumeGomez** [comment]
Should be good now. :'(

**Comment by GuillaumeGomez** [comment]
I fixed tidy issue.

**Comment by Manishearth** [comment]
@bors-servo try
Looks like it is causing a failure? https://buildbot.rust-lang.org/builders/auto-mac-64-opt-rustbuild/builds/2380

**Comment by TimNN** [comment]
The problem is that [this test](https://github.com/rust-lang/rust/pull/36102/files#diff-ddc08ee4c61a0e97fbf38d275274b05eR33) passes on OS X (since the tests are apparently compiled as libraries, so no linking happens).

**Comment by GuillaumeGomez** [comment]
Once https://github.com/rust-lang/rust/pull/36320 is merged, it'll be easier to see from where the problem comes.

**Comment by TimNN** [comment]
@GuillaumeGomez: have you seen my comment above?

**Comment by GuillaumeGomez** [comment]
@TimNN: Yes, however it can be the one which fails the tests since it isn't run.

**Comment by GuillaumeGomez** [comment]
I have set the failing test into "ignore mode".

**Comment by jonathandturner** [comment]
@GuillaumeGomez - please ignore the test only for OS X, if that's the troublesome platform:

```rust
// ignore-macos
```

**Comment by TimNN** [comment]
@jonathandturner: Is that possible outside of `compiletest`?

**Comment by jonathandturner** [comment]
@TimNN - yeah it may not work outside of that.  @alexcrichton mentioned it had limitations on irc.

**Comment by jonathandturner** [comment]
If it doesn't support it in the part of code you're in, I guess it's fine to just ignore the whole thing.

**Comment by GuillaumeGomez** [comment]
I'll work on adding later on.

**Comment by GuillaumeGomez** [comment]
Any news?

**Comment by jonathandturner** [comment]
@GuillaumeGomez - if you can't ignore just for the platform where you are in the code, I'm okay with ignoring the test.

**Comment by GuillaumeGomez** [comment]
`rustdoc` doesn't seem to provide such a thing but maybe I missed it?

cc @rust-lang/tools

**Comment by TimNN** [comment]
@GuillaumeGomez: I don't think `rustdoc` provides something like this either, so maybe the best solution is to ignore the test in the explain message and just add a test to `compile-fail`, where `// ignore-macos` can be used as far as I now.

(I just noticed that there is already a similar test for the compiler flag, linking to a Framework: https://github.com/rust-lang/rust/blob/master/src/test/compile-fail/manual-link-framework.rs).

**Comment by GuillaumeGomez** [comment]
It just fails without giving an error code. So I guess the better should be to keep it like this.

**Pull request merged**